core/controllers/editor.py

```diff
@@ -924,3 +945,37 @@ class StartedTutorialEventHandler(EditorHandler):
     def post(self):
         """Handles GET requests."""
         user_services.record_user_started_state_editor_tutorial(self.user_id)
+
+
+class EditorAutosaveHandler(ExplorationHandler):
+    """Handles requests from the editor for draft autosave."""
+
+    @require_editor
+    def put(self, exploration_id):
+        """Handles PUT requests for draft updation."""
+        # Raise an Exception if the draft change list fails non-strict
+        # validation.
+        try:
+            change_list = self.payload.get('change_list')
+            version = self.payload.get('version')
+            exp_services.create_or_update_draft(
+                exploration_id, self.user_id, change_list, version,
+                datetime.datetime.utcnow())
+        except utils.ValidationError as e:
+            # If this exception is raised, let the user know that their changes
+            # have been auto discarded.
+            exp_services.discard_draft(exploration_id, self.user_id)
+            raise self.InvalidInputException(e)
+
+        # If the value passed here is False, have the user discard the draft
+        # changes. We save the draft to the datastore even if the version is
+        # invalid, so that it is available for recovery later.
+        self.render_json({
+            'is_version_of_draft_valid': exp_services.is_version_of_draft_valid(
+                exploration_id, version)})
```

Comment:
I believe in this case the front-end should have just sent a copy of the draft changes to the backend (the change_list here) so it should still have a copy itself that it can use.